writer.go

```diff
@@ -141,31 +138,56 @@ func (w *writeWorker) done(reqs []*request, err error) {
 	}
 }
 
-func (w *writeWorker) writeToLSM(b *request) error {
-	if len(b.Ptrs) != len(b.Entries) {
-		return errors.Errorf("Ptrs and Entries don't match: %+v", b)
+func (w *writeWorker) writeToLSM(entries []*Entry, ptrs []valuePointer) error {
+	if len(ptrs) != len(entries) {
+		return errors.Errorf("Ptrs and Entries don't match: %v, %v", entries, ptrs)
+	}
+
+	var i uint64
+	var err error
+	for err = w.ensureRoomForWrite(); err == errNoRoom; err = w.ensureRoomForWrite() {
+		i++
+		if i%100 == 0 {
+			log.Warnf("Making room for writes")
+		}
+		// We need to poll a bit because both hasRoomForWrite and the flusher need access to s.imm.
+		// When flushChan is full and you are blocked there, and the flusher is trying to update s.imm,
+		// you will get a deadlock.
+		time.Sleep(10 * time.Millisecond)
+	}
+	if err != nil {
+		return err
 	}
 
-	for i, entry := range b.Entries {
+	es := make([]table.Entry, 0, len(entries)-1)
+	for i, entry := range entries {
 		if entry.meta&bitFinTxn != 0 {
 			continue
 		}
 		if w.shouldWriteValueToLSM(entry) { // Will include deletion / tombstone case.
-			w.mt.Put(entry.Key,
-				y.ValueStruct{
+			es = append(es, table.Entry{
+				Key: entry.Key,
+				Value: y.ValueStruct{
 					Value:    entry.Value,
 					Meta:     entry.meta,
 					UserMeta: entry.UserMeta,
-				})
+					Version:  y.ParseTs(entry.Key),
+				},
+			})
 		} else {
 			var offsetBuf [vptrSize]byte
-			w.mt.Put(entry.Key,
-				y.ValueStruct{
-					Value:    b.Ptrs[i].Encode(offsetBuf[:]),
+			es = append(es, table.Entry{
+				Key: entry.Key,
+				Value: y.ValueStruct{
+					Value:    ptrs[i].Encode(offsetBuf[:]),
```

Comment:
The `offsetBuf` is on the stack, it's OK for the old code because SkipList copies the value, but here we put it into a slice, the data is not copied.

Reply:
A workaround for it needs some refactor to whole codebase.

Add an additional field `inlineBuf [vptrSize]byte` into `y.ValueStruct`. We can `copy` `offsetBuf` into here to avoid head allocation.

Then introduce
```go
func (v *ValueStruct) Value() []byte {
     if v.value == nil {
       return v.inlineBuf[:]
     }
     return v.value
}
```

The overhead is the size of `y.ValueStruct` is increased

Reply:
We can reuse the `entry.Value` to encode the value pointer.

Reply:
I think it is okay to keep this at the moment, and then optimize it if there is a problem.

Reply:
Good idea👍